sdks/python/apache_beam/transforms/combiners.py

```diff
@@ -193,9 +193,13 @@ def Of(pcoll, n, compare=None, *args, **kwargs):
           original_compare = compare
           compare = lambda a, b: original_compare(b, a)
       # This is a more efficient global algorithm.
+      top_per_bundle = pcoll | core.ParDo(_TopPerBundle(n, compare))
+      # If pcoll is empty, we can't guerentee that top_per_bundle
+      # won't be empty, so inject at least one empty accumulator
+      # so that downstream is guerenteed to produce non-empty output.
+      empty_bundle = pcoll.pipeline | core.Create([(None, [])])
       return (
-          pcoll
-          | core.ParDo(_TopPerBundle(n, compare))
+          (top_per_bundle, empty_bundle) | core.Flatten()
```

Comment:
This is nice, since the Flatten is before GBK and there won't be a need to materialize the entire results of Top in a single record (at least for some runners), which increases the maximum realizable Top (since we wouldn't just be limited on what fits in a single record, but only on what fits in a single record, eg 2GB).

Might we be able to do something similar for CombineGlobally's with_defaults which is the default?
I think as implemented today, that one has the limit of single-record (2GB) as opposed to being limited by the RAM of a machine (which is often a lot more than 2GB).

Reply:
Yes, it also provides more fusion opportunities. I was also thinking about how to apply this to CombineGlobally in general. It's tricky because the empty combiner needs to be flattened with an intermediate collection that is created after combiner lifting, and I haven't figured out how to express this cleanly (e.g. have a "must provide output for key" section on combine per key?)

Reply:
Agreed re fusion as well.

Idea: Perhaps if CoGroupByKey allowed specification of Combiners per collection then we could realize this on top of it?

I guess no good way to address this in the current PR though.